# Load all datasets

In [1]:
from scipy.io import arff
import pandas as pd

# Glass
data_glass, meta_glass = arff.loadarff("Glass/Glass_withoutdupl_norm.arff")
df_glass = pd.DataFrame(data_glass)

# ALOI
data_aloi, meta_aloi = arff.loadarff("ALOI/ALOI_withoutdupl_norm.arff")
df_aloi = pd.DataFrame(data_aloi)

# Ionosphere
data_ionosphere, meta_ionosphere = arff.loadarff("Ionosphere/Ionosphere_withoutdupl_norm.arff")
df_ionosphere = pd.DataFrame(data_ionosphere)

# Understand all data

## Glass Dataset

In [2]:
df_glass.head()

,var_0000,var_0001,var_0002,var_0003,var_0004,var_0005,var_0006,id,outlier
0,0.467651,0.321584,0.768880,0.246630,0.838799,0.099737,0.298340,1.0,b'no'
1,0.496412,0.220491,0.776032,0.316598,0.919973,0.089145,0.279479,2.0,b'no'
2,0.519133,0.404464,0.768012,0.334978,0.801622,0.092369,0.271238,3.0,b'no'
3,0.199650,0.547373,0.374284,0.362223,0.817017,0.000000,0.177913,4.0,b'yes'
4,0.847261,0.286361,0.000000,0.217792,0.000000,0.019135,1.000000,5.0,b'no'


The dataset contains seven numerical features describing each data point (from var_0000 to var_0006).

Each observation is assigned a unique identifier ranging from 1 to 214.

The final column, outlier, indicates whether a data point is considered an outlier and serves as the ground truth label for evaluation.

In [3]:
df_glass.shape

(214, 9)

The dataset consists of 214 data points and 9 columns.

In [4]:
df_glass['outlier'].value_counts()

outlier
b'no'     205
b'yes'      9
Name: count, dtype: int64

The dataset contains 9 outliers and 205 inliers, which matches the statistics reported on the DAMI webpage for the Glass dataset.

## ALOI Dataset

In [5]:
df_aloi.head()

,att1,att2,att3,att4,att5,att6,att7,att8,att9,att10,...,att20,att21,att22,att23,att24,att25,att26,att27,outlier,id
0,0.784999,0.000018,0.0,0.000093,0.0,0.0,0.0,0.0,0.0,0.139811,...,0.001189,0.0,0.115728,0.023449,0.0002,0.000220,0.034952,0.046914,b'yes',1.0
1,0.958088,0.000000,0.0,0.001671,0.0,0.0,0.0,0.0,0.0,0.019556,...,0.000000,0.0,0.022035,0.007516,0.0000,0.000000,0.001033,0.000000,b'yes',2.0
2,0.938768,0.000000,0.0,0.005146,0.0,0.0,0.0,0.0,0.0,0.018451,...,0.000000,0.0,0.035542,0.011982,0.0000,0.001595,0.019520,0.000000,b'yes',3.0
3,0.954775,0.000000,0.0,0.001427,0.0,0.0,0.0,0.0,0.0,0.024944,...,0.000000,0.0,0.019941,0.000805,0.0000,0.000000,0.000035,0.000000,b'yes',4.0
4,0.933601,0.000000,0.0,0.001682,0.0,0.0,0.0,0.0,0.0,0.037002,...,0.000000,0.0,0.046759,0.002663,0.0000,0.000339,0.001359,0.000000,b'yes',5.0


Compared to the Glass dataset, ALOI contains a significantly larger number of features. Each data point is described by 27 numerical attributes.

The last two columns represent the outlier label and the corresponding unique identifier.

In [6]:
df_aloi.shape

(49534, 29)

This dataset contains 49'534 data points, which is substantially larger than the Glass dataset.

In [7]:
df_aloi['outlier'].value_counts()

outlier
b'no'     48026
b'yes'     1508
Name: count, dtype: int64

The dataset includes 1'508 outliers and 48'026 inliers, which corresponds to the distribution reported on the DAMI webpage.

## Ionosphere dataset

In [8]:
df_ionosphere.head()

,var_0000,var_0001,var_0002,var_0003,var_0004,var_0005,var_0006,var_0007,var_0008,var_0009,...,var_0024,var_0025,var_0026,var_0027,var_0028,var_0029,var_0030,var_0031,id,outlier
0,0.997695,0.470555,0.926215,0.511530,0.91699,0.311460,1.000000,0.518800,0.926215,0.411225,...,0.705390,0.269160,0.606330,0.329550,0.711335,0.227565,0.593205,0.273500,1.0,b'no'
1,1.000000,0.405855,0.965175,0.319220,0.44566,0.032015,1.000000,0.477255,0.754370,0.161285,...,0.397660,0.407995,0.404800,0.442035,0.416870,0.468560,0.431310,0.487765,2.0,b'yes'
2,1.000000,0.483175,1.000000,0.502425,1.00000,0.439690,0.944825,0.505990,0.865410,0.526730,...,0.794920,0.389275,0.715500,0.413175,0.802180,0.379100,0.780225,0.308810,3.0,b'no'
3,1.000000,0.274195,1.000000,1.000000,0.85608,0.000000,0.500000,0.500000,0.500000,0.500000,...,0.758065,1.000000,1.000000,0.399505,0.628410,1.000000,0.338090,1.000000,4.0,b'yes'
4,1.000000,0.487995,0.970700,0.532655,0.96053,0.383725,0.885760,0.418005,0.763990,0.398625,...,0.566450,0.233970,0.512155,0.189015,0.471465,0.202135,0.476960,0.171515,5.0,b'no'


The dataset contains 31 numerical feature columns describing each data point. The remaining two columns correspond to the outlier label and the unique identifier.

In [9]:
df_ionosphere.shape

(351, 34)

The dataset consists of 351 data points, which is considerably smaller than the ALOI dataset.

In [10]:
df_ionosphere['outlier'].value_counts()

outlier
b'no'     225
b'yes'    126
Name: count, dtype: int64

This dataset presents a more balanced distribution between outliers and inliers. It contains 126 outliers (35.9%) and 225 inliers (64.1%), which aligns with the statistics reported on the website.

# Setup LOF Algorithm

In [11]:
# Create a ground_truth for each dataset
y_true_glass = df_glass['outlier'].map({b'no': 0, b'yes': 1})
y_true_aloi = df_aloi['outlier'].map({b'no': 0, b'yes': 1})
y_true_ionosphere = df_ionosphere['outlier'].map({b'no': 0, b'yes': 1})

# Drop the column to remove data leakage
X_glass = df_glass.drop(columns=['outlier', 'id'])
X_aloi = df_aloi.drop(columns=['outlier', 'id'])
X_ionosphere = df_ionosphere.drop(columns=['outlier', 'id'])

# Search for the Best Contamination / n_neighbors Combination (Maximizing F1-Score)

In this section, we perform an exhaustive search over different values of `contamination` and `n_neighbors` in order to find the combination that maximizes the F1-score.

⚠️ Note: Running this search is not mandatory. You may skip the execution and simply refer to the reported results shown below.

In [12]:
from sklearn.neighbors import LocalOutlierFactor
from sklearn.metrics import f1_score
import numpy as np

def search_best_lof(X, y_true, n_neighbors_list, contamination_list):
    best_f1 = -1 
    best_params = {}
    
    for n_neighbors in n_neighbors_list:
        for contamination in contamination_list:
            
            clf = LocalOutlierFactor(
                n_neighbors=n_neighbors,
                contamination=contamination
            )
            
            y_pred = clf.fit_predict(X)
            y_pred_converted = (y_pred == -1).astype(int)
            
            f1 = f1_score(y_true, y_pred_converted)
            
            if f1 > best_f1:
                best_f1 = f1
                best_params = {
                    'n_neighbors': n_neighbors,
                    'contamination': contamination
                }
    
    return best_params, best_f1


# Neighbor ranges
n_neighbors_list_glass = [2,3,4,5,6,7,8,10,15,20]
n_neighbors_list_aloi = [2,3,4,5,6,7,8,9,10]
n_neighbors_list_ionosphere = [2,3,4,5,6,7,8,10,15,20]


# Realistic contamination ranges
contamination_list_glass = np.arange(0.02, 0.081, 0.005) 
contamination_list_aloi = np.arange(0.01, 0.061, 0.005)
contamination_list_ionosphere = np.arange(0.30, 0.451, 0.01)


# Run searches
best_params_glass, best_f1_glass = search_best_lof(
    X_glass, y_true_glass,
    n_neighbors_list_glass,
    contamination_list_glass
)

best_params_aloi, best_f1_aloi = search_best_lof(
    X_aloi, y_true_aloi,
    n_neighbors_list_aloi,
    contamination_list_aloi
)

best_params_ionosphere, best_f1_ionosphere = search_best_lof(
    X_ionosphere, y_true_ionosphere,
    n_neighbors_list_ionosphere,
    contamination_list_ionosphere
)

print("Glass:", best_params_glass, best_f1_glass)
print("ALOI:", best_params_aloi, best_f1_aloi)
print("Ionosphere:", best_params_ionosphere, best_f1_ionosphere)

Glass: {'n_neighbors': 2, 'contamination': np.float64(0.05)} 0.4
ALOI: {'n_neighbors': 5, 'contamination': np.float64(0.05499999999999999)} 0.21403260099220411
Ionosphere: {'n_neighbors': 8, 'contamination': np.float64(0.38000000000000006)} 0.833976833976834


## Best Combinations

k = number of neighbors  
cont = contamination  

Glass      : k = 2 and cont = 0.05  
ALOI       : k = 5 and cont = 0.055  
Ionosphere : k = 8 and cont = 0.38  

To obtain these results, we performed a grid search over predefined ranges of `n_neighbors` and `contamination`.

The tested values were not arbitrary. We chose realistic ranges for each dataset:

- The n_neighbors values were selected to explore small to moderate neighborhood sizes.
- The contamination ranges were centered around the expected proportion of outliers in each dataset (approximately 4% for Glass, 3% for ALOI, and 36% for Ionosphere).

By restricting the search to realistic contamination intervals, we avoid testing implausible combinations and we avoid excessive computational time

## LOF Results Summary

The Local Outlier Factor (LOF) algorithm was evaluated on the three datasets using the selected hyperparameters above.

- **Glass**: The model produced 12 classification errors, with a precision of 0.364, a recall of 0.444, and an F1-score of 0.4. This indicates moderate detection performance, with slightly better recall than precision.

- **ALOI**: The model resulted in 3327 errors. Precision was 0.166, recall was 0.3, and the F1-score was 0.214. Performance is relatively low, suggesting difficulty in accurately identifying outliers in this dataset.

- **Ionosphere**: The model achieved 43 errors, with a precision of 0.812, recall of 0.857, and an F1-score of 0.834. This reflects strong outlier detection performance, with both high precision and recall.

We can conclude that LOF performs best on the Ionosphere dataset.

In [21]:
from sklearn.neighbors import LocalOutlierFactor
from sklearn.metrics import f1_score, precision_score, recall_score


# ---------- GLASS ----------
clf_glass = LocalOutlierFactor(
    # Combination from previous result!
    n_neighbors=2,
    contamination=0.05
)

y_pred_glass = clf_glass.fit_predict(X_glass)
y_pred_glass_conv = (y_pred_glass == -1).astype(int)

n_errors_glass = (y_pred_glass_conv != y_true_glass).sum()

print("Glass dataset errors:", n_errors_glass)
print("Precision:", precision_score(y_true_glass, y_pred_glass_conv))
print("Recall:", recall_score(y_true_glass, y_pred_glass_conv))
print("F1:", f1_score(y_true_glass, y_pred_glass_conv))


# ---------- ALOI ----------
clf_aloi = LocalOutlierFactor(
    # Combination from previous result!
    n_neighbors=5,
    contamination=0.055
)

y_pred_aloi = clf_aloi.fit_predict(X_aloi)
y_pred_aloi_conv = (y_pred_aloi == -1).astype(int)

n_errors_aloi = (y_pred_aloi_conv != y_true_aloi).sum()

print("\nALOI dataset errors:", n_errors_aloi)
print("Precision:", precision_score(y_true_aloi, y_pred_aloi_conv))
print("Recall:", recall_score(y_true_aloi, y_pred_aloi_conv))
print("F1:", f1_score(y_true_aloi, y_pred_aloi_conv))


# ---------- IONOSPHERE ----------
clf_ionosphere = LocalOutlierFactor(
    # Combination from previous result!
    n_neighbors=8,
    contamination=0.378
)

y_pred_ionosphere = clf_ionosphere.fit_predict(X_ionosphere)
y_pred_ionosphere_conv = (y_pred_ionosphere == -1).astype(int)

n_errors_ionosphere = (y_pred_ionosphere_conv != y_true_ionosphere).sum()

print("\nIonosphere dataset errors:", n_errors_ionosphere)
print("Precision:", precision_score(y_true_ionosphere, y_pred_ionosphere_conv))
print("Recall:", recall_score(y_true_ionosphere, y_pred_ionosphere_conv))
print("F1:", f1_score(y_true_ionosphere, y_pred_ionosphere_conv))

Glass dataset errors: 12
Precision: 0.36363636363636365
Recall: 0.4444444444444444
F1: 0.4

ALOI dataset errors: 3327
Precision: 0.16623853211009174
Recall: 0.3003978779840849
F1: 0.21403260099220411

Ionosphere dataset errors: 43
Precision: 0.8120300751879699
Recall: 0.8571428571428571
F1: 0.833976833976834


# Conclusion with the DAMI Website Results

We now compare our results with those reported on the DAMI website for the LOF algorithm.

- **Glass**: The best reported F1-score is 0.375, which is slightly lower than our result. This difference may be explained by the fact that the DAMI evaluation likely used a fixed contamination value, whereas we optimized the hyperparameters within a plausible range.

- **ALOI**: The reported best F1-score is 0.217, while we obtained 0.214. The results are very close, with DAMI achieving a slightly better performance.

- **Ionosphere**: The reported best F1-score is 0.83, compared to our 0.812. The performance is again very similar, with a small difference in favor of the DAMI result.

Overall, our results are consistent with the benchmark values.




>It is important to note that LOF is an unsupervised algorithm. 
However, by tuning hyperparameters using ground truth labels to maximize F1-score, 
we effectively introduce supervised model selection. 
This may slightly inflate performance compared to a fully unsupervised setting.
